### PR TITLE
Loop rules

### PR DESCRIPTION
**Pull request by schalkms**

### CODE REVIEW

**Comment by arturbosch** [comment]
Please rebase your branch to resolve merge conflicts.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.psi.KtConstantExpression
+import org.jetbrains.kotlin.psi.KtForExpression
+
+class InvalidLoopCondition(config: Config = Config.empty) : Rule(config) {
+
+	override val issue = Issue(javaClass.simpleName,
+			Severity.Defect,
+			"If a for loops condition is false before the first iteration, the loop will never get executed.",
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt]: `s/If/When`?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	private fun hasInvalidLoopRange(range: Array<PsiElement>): Boolean {
+		val lowerValue = getIntValueForElement(range[0])
+		val upperValue = getIntValueForElement(range[2])
+		if (lowerValue == null || upperValue == null) {
+			return false
+		}
+		return when (range[1].text) {
+			".." -> checkRangeTo(lowerValue, upperValue)
+			"downTo" -> checkDownTo(lowerValue, upperValue)
+			"until" -> checkUntil(lowerValue, upperValue)
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt]: does this also handle the `step` cases or `reversed`?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+			report(CodeSmell(issue, Entity.from(loopExpression)))
+		}
+		super.visitLoopExpression(loopExpression)
+	}
+
+	private fun countBreakAndReturnStatements(body: KtExpression?): Int {
+		if (body == null) {
+			return 0
+		}
+		return body.countBreakAndReturnStatementsInLoop()
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt]: `return return body?.countBreakAndReturnStatementsInLoop() ?: 0` ?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/KT.kt [hunk 4]
```diff
@@ -19,8 +19,10 @@ enum class Case(val file: String) {
 	Exceptions("/cases/Exceptions.kt"),
 	ExceptionRaisedInMethods("/cases/ExceptionRaisedInMethods.kt"),
 	FinalClass("/cases/FinalClass.kt"),
+	InvalidLoopCondition("/cases/InvalidLoopCondition.kt"),
 	IteratorImpl("/cases/IteratorImpl.kt"),
-	LabeledExpression("/cases/LabeledExpression.kt"),
+  LoopWithTooManyJumpStatements("cases/LoopWithTooManyJumpStatements.kt"),
```
Comment (vanniktech) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/KT.kt]: indent is off

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt (hunk 1 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt]: Thanks for your review. 
Well, this only works for loops. Sure it is possible to detect always true or false condition for conditionals, but this is out of the scope of this rule.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt (hunk 2 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt]: If the range is invalid `step` does not make any difference. The same is true for `reversed`.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/KT.kt (hunk 4 above)
Comment (schalkms) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/KT.kt]: Good catch! Thanks.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt (hunk 3 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt]: Yes. I can use the shorter syntax.

File: detekt-rules/src/test/resources/cases/InvalidLoopCondition.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package cases
+
+@Suppress("unused")
+fun loopConditions() {
+	for (i in 2..2) { }
+	for (i in 2..1) { } // violation
```
Comment (arturbosch) [inline detekt-rules/src/test/resources/cases/InvalidLoopCondition.kt]: Suppress empty range here

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+			else -> false
+		}
+	}
+
+	private fun getIntValueForElement(element: PsiElement): Int? {
+		val constant = element as? KtConstantExpression ?: return null
+		return try {
+			Integer.parseInt(constant.text)
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/InvalidLoopCondition.kt]: You can use `contant.text.toIntOrNull()` here

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.jetbrains.kotlin.psi.KtLoopExpression
+
+class LoopWithTooManyJumpStatements(config: Config = Config.empty) : Rule(config) {
+
+	override val issue = Issue(javaClass.simpleName, Severity.Style,
+			"The loop contains more than one break or continue statement. " +
+					"The code should be refactored to increase readability.", Debt.TEN_MINS)
+
+	override fun visitLoopExpression(loopExpression: KtLoopExpression) {
+		if (countBreakAndReturnStatements(loopExpression.body) > 1) {
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatements.kt]: We should have this configurable like everything in detekt ?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatementsSpec.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import org.jetbrains.spek.api.dsl.given
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.subject.SubjectSpek
+
+class LoopWithTooManyJumpStatementsSpec : SubjectSpek<LoopWithTooManyJumpStatements>({
+	subject { LoopWithTooManyJumpStatements() }
+
+	given("loops with multiple break or continue statements") {
+
+		it("reports loops with more than 1 break or continue statement") {
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/LoopWithTooManyJumpStatementsSpec.kt]: After making the rule configurable, an extra test case would be nice :)

File: detekt-rules/src/test/resources/cases/LoopWithTooManyJumpStatements.kt [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package cases
+
+@Suppress("unused")
+class LoopWithTooManyJumpStatements {
+
+	fun tooManyJumps() { // reports 3
+		val i = 0
+		for (j in 1..2) {
```
Comment (arturbosch) [inline detekt-rules/src/test/resources/cases/LoopWithTooManyJumpStatements.kt]: Suppress all intellij inspections here too

File: detekt-rules/src/test/resources/cases/UnconditionalJumpStatementInLoop.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package cases
+
+@Suppress("unused")
+class UnconditionalJumpStatementInLoop {
+
+	fun jumpStatementsInLoop() { // reports 6
+		for (i in 1..2) break
+		for (i in 1..2) continue
```
Comment (arturbosch) [inline detekt-rules/src/test/resources/cases/UnconditionalJumpStatementInLoop.kt]: Just in case, also here everything suppressing on class level

**Comment by arturbosch** [comment]
Nice rules! More bug rules are always nice for promoting detekt! Have you tried rebasing your branch? If GitHub can't merge this, it will make much more work for me. `git merge master` into your branch and `git rebase` will make you resolve the conflict, big thanks!

**Comment by schalkms** [comment]
@arturbosch Thanks for the review. I've done a merge and rebase. Github doesn't report conflicts.

**Comment by arturbosch** [comment]
It was still conflicted for me in the file `KT`. Just merged it manually :), closing.

**Pull request closed without merging**